### ISSUE TITLE
Memory leak in `cosmic-settings`

### ISSUE THREAD

**Issue by andre4ik3** [pop-os/cosmic-settings#843]
After exiting from COSMIC, there were still 3 instances of `cosmic-settings` left running (from me opening & closing it from within COSMIC), then while I was in KDE Plasma, the `cosmic-settings` instances were leaking memory in the background, and two were killed by the kernel OOM killer: [Kernel OOM logs](https://github.com/user-attachments/files/18255626/kernel-oom.txt). The third one I recorded on video, it went up to around 94 GiB and then hovered there. However it would not respond to quit requests, I had to send a SIGKILL for it to close, even after it stopped growing in memory usage.

https://github.com/user-attachments/assets/ff16f73f-9c93-4b4d-a722-101a423a33bb

After I killed it, it sent this in journal:

```
app-cosmic-com.system76.CosmicAppList-344252.scope: Consumed 7min 19.390s CPU time, 94.1G memory peak.
```

OS: latest NixOS unstable, using latest nixos-cosmic flake
Version: `cosmic-settings-1.0.0-alpha.4-unstable-2024-12-17`

**Comment by bhh32** [comment]
Do you have any steps for us to try and replicate the issue? If we can get it to happen again, it may be easier for us to track down where in the code this is stemming from and fix it.

**Comment by andre4ik3** [comment]
I was able to reproduce it just now again on latest 2025.01.02 version. Each cosmic-setting process is using ~25 GiB, for a total of ~100 GiB.

![изображение](https://github.com/user-attachments/assets/5232bef0-591d-4079-9980-46cb2d2902cf)
![изображение](https://github.com/user-attachments/assets/03d37b00-a936-44a4-bae7-02adff937874)

1. Go to COSMIC (my display-manager is SDDM, not sure if it's relevant)
2. Open cosmic-settings, go to "Monitors" section
3. If you want, close and open a few times, it will reopen on "Monitors" section (in my screenshots I closed and reopened 4 times)
4. Exit COSMIC desktop
5. Enter KDE plasma, open system monitor and watch RAM go up (goes up faster the more times you opened/closed cosmic-settings)
6. (Optional) Kill all the cosmic-settings processes quickly before the system OOM's :)

**Comment by bhh32** [comment]
> I was able to reproduce it just now again on latest 2025.01.02 version. Each cosmic-setting process is using ~25 GiB, for a total of ~100 GiB.
> 
> 
> 
> ![изображение](https://github.com/user-attachments/assets/5232bef0-591d-4079-9980-46cb2d2902cf)
> 
> ![изображение](https://github.com/user-attachments/assets/03d37b00-a936-44a4-bae7-02adff937874)
> 
> 
> 
> 1. Go to COSMIC (my display-manager is SDDM, not sure if it's relevant)
> 
> 2. Open cosmic-settings, go to "Monitors" section
> 
> 3. If you want, close and open a few times, it will reopen on "Monitors" section (in my screenshots I closed and reopened 4 times)
> 
> 4. Exit COSMIC desktop
> 
> 5. Enter KDE plasma, open system monitor and watch RAM go up (goes up faster the more times you opened/closed cosmic-settings)
> 
> 6. (Optional) Kill all the cosmic-settings processes quickly before the system OOM's :)

When you switched from COSMIC to KDE, did you log out and then back in? If not, how are you switching DEs? 
  
It could be related to you running SSDM, but I don't know that for sure. Is there a way for you to check if this happens when you're running cosmic-comp instead of SSDM?

**Comment by andre4ik3** [comment]
> When you switched from COSMIC to KDE, did you log out and then back in? If not, how are you switching DEs?

Yes, I logged out from COSMIC, switched the session in SDDM from COSMIC to Plasma (Wayland), then logged back in.

> It could be related to you running SSDM, but I don't know that for sure. Is there a way for you to check if this happens when you're running cosmic-comp instead of SSDM?

Just checked by doing `services.displayManager.cosmic-greeter.enable = true;`, and following the steps outlined above, the issue remains. I opened and closed cosmic-settings 3 times, and there was 3 cosmic-settings leaking memory after logging out from COSMIC and logging into KDE.

**Comment by bhh32** [comment]
> > When you switched from COSMIC to KDE, did you log out and then back in? If not, how are you switching DEs?
> 
> 
> 
> Yes, I logged out from COSMIC, switched the session in SDDM from COSMIC to Plasma (Wayland), then logged back in.
> 
> 
> 
> > It could be related to you running SSDM, but I don't know that for sure. Is there a way for you to check if this happens when you're running cosmic-comp instead of SSDM?
> 
> 
> 
> Just checked by doing `services.displayManager.cosmic-greeter.enable = true;`, and following the steps outlined above, the issue remains. I opened and closed cosmic-settings 3 times, and there was 3 cosmic-settings leaking memory after logging out from COSMIC and logging into KDE.

Ok, thanks for checking and giving all the information.

**Comment by aurabindo** [comment]
Also seeing a memleak. I'm on Cachy OS, and also have rusty userspace scheduler in use.
Dmesg: https://pastebin.com/k7b9jN4b

**Comment by MalcolmReed-ent** [comment]
Hey guys, I discovered a memory leak while running Alpha 5 on Arch Linux. I encountered it through two different scenarios: first while adjusting display settings (though I didn't track the exact steps), and second by running cosmic-settings, logging out, and then launching settings again, the second occurrence was the one I gathered info from. 

I ran various versions of valgrind here is the first

`valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes --verbose --log-file=cosmic-settings-valgrind.log cosmic-settings`

and the second was

`valgrind --tool=massif cosmic-settings 2> cosmic-settings-memory-profile.log`

and here are the various files i have accumulated from it


[massif.out.172478.log](https://github.com/user-attachments/files/18391133/massif.out.172478.log)

[cosmic-settings-memory-profile.log](https://github.com/user-attachments/files/18391130/cosmic-settings-memory-profile.log)

[cosmic-settings-valgrind.log](https://github.com/user-attachments/files/18391131/cosmic-settings-valgrind.log)

**Comment by bhh32** [comment]
So it seems the 3 examples we currently have of this are all coming from non-Pop!_OS distros. Would that be correct? I need to do some more testing on Pop!_OS to see if this is something that can be traced to outside distros.

**Comment by bhh32** [comment]
> Also seeing a memleak. I'm on Cachy OS, and also have rusty userspace scheduler in use.
> 
> Dmesg: https://pastebin.com/k7b9jN4b
> 
> 
> 
> I was changing changing some keyboard shortcuts, and changing various fonts when I noticed the hang.

Do you also have KDE installed?

**Comment by bhh32** [comment]
> Hey guys, I discovered a memory leak while running Alpha 5 on Arch Linux. I encountered it through two different scenarios: first while adjusting display settings (though I didn't track the exact steps), and second by running cosmic-settings, logging out, and then launching settings again, the second occurrence was the one I gathered info from. 
> 
> 
> 
> I ran various versions of valgrind here is the first
> 
> 
> 
> `valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes --verbose --log-file=cosmic-settings-valgrind.log cosmic-settings`
> 
> 
> 
> and the second was
> 
> 
> 
> `valgrind --tool=massif cosmic-settings 2> cosmic-settings-memory-profile.log`
> 
> 
> 
> and here are the various files i have accumulated from it
> 
> 
> 
> 
> 
> [massif.out.172478.log](https://github.com/user-attachments/files/18391133/massif.out.172478.log)
> 
> 
> 
> [cosmic-settings-memory-profile.log](https://github.com/user-attachments/files/18391130/cosmic-settings-memory-profile.log)
> 
> 
> 
> [cosmic-settings-valgrind.log](https://github.com/user-attachments/files/18391131/cosmic-settings-valgrind.log)
> 
> 

Do you also have KDE and/or SSDM installed?

**Comment by MalcolmReed-ent** [comment]
> > Hey guys, I discovered a memory leak while running Alpha 5 on Arch Linux. I encountered it through two different scenarios: first while adjusting display settings (though I didn't track the exact steps), and second by running cosmic-settings, logging out, and then launching settings again, the second occurrence was the one I gathered info from.
> > I ran various versions of valgrind here is the first
> > `valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes --verbose --log-file=cosmic-settings-valgrind.log cosmic-settings`
> > and the second was
> > `valgrind --tool=massif cosmic-settings 2> cosmic-settings-memory-profile.log`
> > and here are the various files i have accumulated from it
> > [massif.out.172478.log](https://github.com/user-attachments/files/18391133/massif.out.172478.log)
> > [cosmic-settings-memory-profile.log](https://github.com/user-attachments/files/18391130/cosmic-settings-memory-profile.log)
> > [cosmic-settings-valgrind.log](https://github.com/user-attachments/files/18391131/cosmic-settings-valgrind.log)
> 
> Do you also have KDE and/or SSDM installed?

No, just cosmic, i removed all of kde and qt hours before running just cosmic

**Comment by mmstick** [comment]
Make sure you have the latest version of cosmic-settings from git. Arch Linux users can use the Chaotic AUR. Font caching issues were resolved last week, as well as a bounded channel on the display settings page blocking the app from exiting.

**Comment by mmstick** [comment]
See https://github.com/pop-os/cosmic-settings/commit/7576ad1af70b92da436720aac34ab114e82ff36c

**Comment by rchennau** [comment]
I was able to reproduce this and trace it to an exact source location. PR #2020 contains a fix — details below for anyone hitting this in the meantime.

## Root cause

`cache_display_image()` in `pages/desktop/wallpaper/mod.rs:448` calls `ImageHandle::from_rgba()` on every `Message::UpdateState` event. That event fires once per wallpaper rotation (inotify watch on the cosmic-bg state file).

`from_rgba` unconditionally calls `Id::unique()`, which is `AtomicU64::fetch_add` — every call gets a globally unique, monotonically-increasing ID. A cache hit is structurally impossible.

```rust
// pop-os/iced core/src/image.rs:216
static NEXT_ID: AtomicU64 = AtomicU64::new(0);
Self(_Id::Unique(NEXT_ID.fetch_add(1, atomic::Ordering::Relaxed)))
```

Each unique ID is inserted into the wgpu raster cache (`wgpu/src/image/raster.rs` — a `FxHashMap<image::Id, Memory>`). Eviction only happens in `trim()` at `raster.rs:78`, which is only reachable from `draw()` at `wgpu/src/lib.rs:162`, which is only called from `present()`. When the Wayland compositor suppresses `RedrawRequested` between rotations — DPMS off, window occlusion, compositor throttling — `present()` never fires, `trim()` never runs, and the map grows unbounded. This is the same gap described in iced-rs/iced#2659, triggered here by the slideshow path.

Observed growth rate: **~9.3 GB/hour** with 93 static wallpapers and a 300-second rotation interval on Pop!_OS 24.04 / cosmic-settings v1.0.12.

## Fix (PR #2020)

Track the `DefaultKey` and fit mode of the last successfully rendered image. At the start of `cache_display_image()`, short-circuit if both match:

```rust
if self.cached_display_handle.is_some()
    && resolved_key == self.cached_display_key
    && Some(self.selected_fit) == self.cached_display_fit
{
    return;
}
```

`from_rgba` is now called only when the wallpaper or fit mode actually changes. RSS stabilises at ~243 MB after the fix vs unbounded growth before.

A complementary fix in `pop-os/iced` — decoupling `trim()` from `present()` — would provide defence-in-depth for similar future issues (iced-rs/iced#2659).

**Comment by andre4ik3** [comment]
@rchennau please don't necropost on an old issue with AI slop :(
